app/renderer/bookmarks.js

```diff
@@ -0,0 +1,323 @@
+                return true
+            }
+            return false
+        })
+    } else {
+        const individualBookmark = storedBookmarkData.filter(
+            e => e.name.replace(specialChars) === input.join(" ")
+                .replace(specialChars))
+        const bookmarksSelectedByTag = storedBookmarkData.filter(
+            e => e.tag.find(t => t === input.join(" ")))
+        const bookmarksSelectedByKeyword = storedBookmarkData.filter(
+            e => e.keywords.find(k => k === input.join("")))
+        selectedBookmarks
+            = individualBookmark.concat(bookmarksSelectedByTag)
+                .concat(bookmarksSelectedByKeyword)
+    }
+    return selectedBookmarks
+}
+
+
+const deleteBookmark = input => {
+    let selectedBookmarks = []
+    if (input.join() === "") {
+        selectedBookmarks = matchBookmarksToInput([`url=${currentPage().src}`])
+    } else {
+        selectedBookmarks = matchBookmarksToInput(input)
+    }
+    notify(`Deleted ${selectedBookmarks.length} bookmarks \
+            that matched the query.`)
+    selectedBookmarks.forEach(sb => {
+        for (let x = 0; x < bookmarkData.bookmarks.length; x++) {
+            if (sb.id === bookmarkData.bookmarks[x].id) {
+                bookmarkData.bookmarks.splice(x, 1)
+            }
+        }
+    })
+    writeBookmarksToFile()
+}
+
+const addTags = tags => {
+    const currentTagIds = bookmarkData.tags.map(tag => tag.id)
+    tags.forEach(t => {
+        if (!currentTagIds.includes(t)) {
+            bookmarkData.tags.push(
+                {
+                    "id": t,
+                    "keywords": [],
+                    "name": ""
+                }
+            )
+        }
+    })
+}
+
+const addFolder = path => {
+    const currentStoredPaths = bookmarkData.folders.map(f => f.path)
+    if (!currentStoredPaths.includes(path)) {
+        bookmarkData.folders.push(
+            {
+                "keywords": [],
+                "name": "",
+                path
+            })
+    }
+}
+
+const getBookmarkData = () => bookmarkData
+
+const writeBookmarksToFile = () => {
+    writeJSON(bookmarksFile, bookmarkData)
+}
+
+const isBookmarkValid = bookmark => {
+    let isValid = true
+    const badOptions = []
+
+    if (!bookmark.name?.trim()) {
+        isValid = false
+        notify("No name provided for the bookmark")
+    }
+
+    // Remove invalid options
+    for (const option in bookmark) {
+        if (!validBookmarkOptions.includes(option)) {
+            badOptions.push(option)
+            delete bookmark[option]
+        }
+    }
+    const names = bookmarkData.bookmarks.map(b => b.name)
+    if (names.includes(bookmark.name)) {
+        isValid = false
+        notify(`A bookmark with name: ${bookmark.name} already exists.`)
+    }
+
+    addTags(bookmark.tag)
+    addFolder(bookmark.path)
+
+    // Check path format
+    // Check color hex values
+    // Check keywords and tags.
+    // Single words?
```

Comment:
I presume these are still on the TODO list? Let me know if you are stuck and/or need help to implement this :)